Develop > Toolkit > Interoperability > XCM-Tools (#158)
polkadot-developers/polkadot-docs: develop/toolkit/interoperability/xcm-tools.md

---
title: XCM Tools
description: Explore essential XCM tools across Polkadot, crafted to enhance cross-chain functionality and integration within the ecosystem.
---

# XCM Tools

> ## Overview

Reviewer: ```suggestion
## Introduction
```

As described in the [Interoperability](/develop/interoperability){target=\_blank} section, XCM (Cross-Consensus Messaging) is a protocol used in the Polkadot and Kusama ecosystems to enable communication and interaction between different chains. It facilitates cross-chain communication, allowing assets, data, and messages to flow seamlessly across the Polkadot ecosystem. XCM plays a key role in fostering interoperability between blockchains within the network.

> As XCM is central to enabling communication between blockchains, developers need robust tools to help interact with, build, and test XCM messages. There are several XCM tools that simplify working with the protocol by providing libraries, frameworks, and utilities that enhance the development process, ensuring that applications built within the Polkadot ecosystem can efficiently use cross-chain functionalities.

Reviewer 2: ```suggestion
As described in the [Interoperability](/develop/interoperability){target=\_blank} section, XCM (Cross-Consensus Messaging) is a protocol used in the Polkadot and Kusama ecosystems to enable communication and interaction between chains. It facilitates cross-chain communication, allowing assets, data, and messages to flow seamlessly across the ecosystem.

As XCM is central to enabling communication between blockchains, developers need robust tools to help interact with, build, and test XCM messages. Several XCM tools simplify working with the protocol by providing libraries, frameworks, and utilities that enhance the development process, ensuring that applications built within the Polkadot ecosystem can efficiently use cross-chain functionalities.

## Popular XCM Tools

### Moonsong Labs XCM Tools

[Moonsong Labs XCM Tools](https://github.com/Moonsong-Labs/xcm-tools){target=\_blank} provides a collection of scripts for managing and testing XCM operations between Polkadot SDK-based runtimes. These tools allow performing tasks like asset registration, channel setup, and XCM initialization. Key features include:

> - **Asset registration** - registers assets, setting units per second (up-front fees) and configuring error (revert) codes

Reviewer: ```suggestion
- **Asset registration** - registers assets, setting units per second (up-front fees), and configuring error (revert) codes
```

> - **Register assets** - sets units per second charges, and configures revert codes

Reviewer 2: Explain this one better. It is difficult to understand what it does

> - **XCM initializer** -  initializes XCM, sets default XCM versions, and configures revert codes for XCM-related precompiles

Reviewer: ```suggestion
- **XCM initializer** - initializes XCM, sets default XCM versions, and configures revert codes for XCM-related precompiles
```

- **Sovereign accounts** - obtain the sovereign account address for any parachain, either on the Relay Chain or for sibling parachains, using a simple command
- **XC20 EVM addresses** - generate XC20-compatible EVM addresses for assets by entering the asset ID, making it easy to integrate assets across EVM-compatible environments
- **Remote accounts** - retrieve remote account addresses needed for multi-location compatibility, using flexible options to specify account types and parachain IDs

To start using these tools, clone the [Astar repository](https://github.com/AstarNetwork/Astar){target=\_blank} and compile the xcm-tools package:

```bash
git clone https://github.com/AstarNetwork/Astar &&
cd Astar &&
cargo build --release -p xcm-tools
```

After compiling, verify the setup with the following command:

```bash
./target/release/xcm-tools --help
```
For more details on using Astar xcm-tools, consult the [official documentation](https://docs.astar.network/docs/learn/interoperability/xcm/integration/tools/){target=\_blank}.

### Chopsticks

> The Chopsticks library provides XCM functionality for testing XCM messages across networks, enabling you to fork multiple parachains along with a relay chain.  For further details, see the [Chopsticks documentation](/develop/application-devs/tooling/chopsticks/overview/#xcm-testing){target=\_blank} about XCM.

Reviewer: ```suggestion
The Chopsticks library provides XCM functionality for testing XCM messages across networks, enabling you to fork multiple parachains along with a relay chain. For further details, see the [Chopsticks documentation](/develop/toolkit/blockchain/fork-live-chains/#xcm-testing){target=\_blank} about XCM.
```
